pkg/resourcevalidator/webhook.go

```diff
@@ -0,0 +1,241 @@
+	decoder                  *admission.Decoder
+	enableResourceValidation bool
+}
+
+// NewShadowPodValidator creates a new shadow pod validator.
+func NewShadowPodValidator(c client.Client, enableResourceValidation bool) *ShadowPodValidator {
+	return &ShadowPodValidator{
+		Client:                   c,
+		PeeringCache:             &peeringCache{ready: false},
+		enableResourceValidation: enableResourceValidation,
+	}
+}
+
+// Handle is the funcion in charge of handling the request.
+// nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	webhooklog.Info(fmt.Sprintf("\t\tOperation: %s", req.Operation))
+
+	if !spv.PeeringCache.ready {
+		webhooklog.Info("PeeringCache not ready")
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("PeeringCache not ready"))
+	}
+
+	var obj runtime.RawExtension
+
+	switch req.Operation {
+	case admissionv1.Create:
+		obj = req.Object
+	case admissionv1.Delete:
+		obj = req.OldObject
+	case admissionv1.Update:
+		return spv.HandleUpdate(req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("unsupported operation %s", req.Operation))
+	}
+
+	// Decode the shadow pod
+	shadowpod, err := spv.DecodeShadowPod(obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	spClusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	// Get ShadowPod Namespace
+	namespace := &corev1.Namespace{}
+	if err := spv.Client.Get(ctx, client.ObjectKey{Name: shadowpod.GetNamespace()}, namespace); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Get Cluster ID origin label of the namespace
+	nsClusterID, found := namespace.Labels[consts.RemoteClusterID]
+	if !found || nsClusterID != spClusterID {
+		webhooklog.Info(fmt.Sprintf("\t\tNamespace Cluster ID [ %s ] does not match the ShadowPod Cluster ID [ %s ]", nsClusterID, spClusterID))
+		return admission.Denied("Namespace Cluster ID label does not match the ShadowPod Cluster ID label")
+	}
+
+	if !spv.enableResourceValidation {
+		return admission.Allowed("")
+	}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		return spv.HandleCreate(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	case admissionv1.Delete:
+		return spv.HandleDelete(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	default:
+		return admission.Denied("Unsupported operation")
+	}
+}
+
+// HandleCreate is the function in charge of handling Creation requests.
+//nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) HandleCreate(ctx context.Context, req admission.Request, shadowpod *vkv1alpha1.ShadowPod,
```

Comment:
Since you are passing the entire req object, dryrun could be extracted internally.